### PR TITLE
NE-551: Configure router to use Power of Two Random Choices with Override

### PR DESCRIPTION
**Pull request by Miciah**
#### Bump openshift/api for unsupportedConfigOverrides

Bump to github.com/openshift/api@0a1e3ae3a9a18c21b34745c0df789fed50d465d5 to get the `spec.unsupportedConfigOverrides` field for the IngressController API.

* `go.mod`: Bump.
* `go.sum`:
* `manifests/00-custom-resource-definition.yaml`:
* `pkg/manifests/bindata.go`:
* `vendor/*`: Regenerate.
* `pkg/operator/controller/ingress/controller.go` (`validateHTTPHeaderBufferValues`):
* `pkg/operator/controller/ingress/controller_test.go` (`TestValidateHTTPHeaderBufferValues`):
* `pkg/operator/controller/ingress/deployment.go` (`desiredRouterDeployment`):
* `pkg/operator/controller/ingress/deployment_test.go` (`TestDesiredRouterDeployment`):
* `test/e2e/http_header_buffer_test.go` (`TestHTTPHeaderBufferSize`): Update to use `spec.tuningOptions`.

#### Configure router to use Power of Two Random Choices

Configure router deployments to use HAProxy's "random" load balancing algorithm, which with its default parameters implements the "Power of Two Random Choices" balancing algorithm.  Allow the user to revert to the previous "Least Connections" balancing algorithm using an unsupported config override.

See https://www.haproxy.com/blog/power-of-two-load-balancing/ for a description and analysis of Power of Two Random Choices.

This change resolves [NE-551](https://issues.redhat.com/browse/NE-551).

* `pkg/operator/controller/ingress/deployment.go` (`RouterLoadBalancingAlgorithmEnvName`): New const.
(`desiredRouterDeployment`): Check `spec.unsupportedConfigOverrides` using the new `ParseUnsupportedConfigOverrides` function, and set `ROUTER_LOAD_BALANCE_ALGORITHM` to "leastconn" if the "loadBalancingAlgorithm=leastconn" override is set.
* `pkg/operator/controller/ingress/deployment_test.go` (`TestDesiredRouterDeployment`): Verify that `desiredRouterDeployment` sets the `ROUTER_LOAD_BALANCE_ALGORITHM` environment variable as expected and correctly handles `spec.unsupportedConfigOverrides`.
* `pkg/util/unsupported_config_overrides.go`: New file. (`ParseUnsupportedConfigOverrides`): New function.  Parse a raw extension value into a map of key-value pairs.
* `pkg/util/unsupported_config_overrides_test.go` (`TestParseUnsupportedConfigOverrides`): New test.  Verify that `ParseUnsupportedConfigOverrides` behaves correctly.
* `test/e2e/operator_test.go` (`availableConditionsForPrivateIngressController`): New variable.  Define the expected status conditions for an ingresscontroller that specifies the "Private" endpoint publishing strategy type and is available.
(`TestTLSSecurityProfile`, `TestRouteAdmissionPolicy`, `TestSyslogLogging`, `TestContainerLogging`, `TestUniqueIdHeader`): Use `availableConditionsForPrivateIngressController`.
(`TestLoadBalancingAlgorithmUnsupportedConfigOverride`): New test.  Verify that the operator configures the router with the "random" load-balancing algorithm by default and allows the algorithm to be changed to "leastconn" algorithm using an unsupported config override.

---

This PR depends on the following PRs:

- [X] https://github.com/openshift/enhancements/pull/664
- [x] https://github.com/openshift/api/pull/878
- [x] https://github.com/openshift/router/pull/192

### CODE REVIEW

**Comment by Miciah** [comment]
The new E2E test failed with the following:
    time="2021-04-03T01:58:59Z" level=info msg="    operator_test.go:1799: failed to update ingresscontroller: json: error calling MarshalJSON for type runtime.RawExtension: invalid character 'l' looking for beginning of value"

So let's see what happens if we use json for `spec.unsupportedConfigOverride`...

**Comment by Miciah** [comment]
Now the verify job is printing nonsense.

/test verify

**Comment by Miciah** [comment]
From [the ingress operator logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/589/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1378177308879228928/artifacts/e2e-aws-operator/gather-must-gather/artifacts/must-gather/namespaces/openshift-ingress-operator/pods/ingress-operator-8457bdd989-l9xsl/ingress-operator/ingress-operator/logs/current.log), it looks like the operator properly handles the unsupported config override.  However, the openshift/api bump bumped some k8s.io packages as well, it looks like controller-runtime is incompatible with the newer k8s.io packages, and I haven't figured out how to deal with that.

**Comment by Miciah** [comment]
I think I've sorted out the dependency mess.  

First, I rebased this PR on https://github.com/openshift/cluster-ingress-operator/pull/523, which bumps controller-runtime to v0.8.3, and then rebased _that_ on the master branch to resolve a merge conflict in `go.sum`.  Unfortunately, that was insufficient:

* Bumping openshift/api required bumping k8s.io packages to v0.21.0-rc.0.
* Bumping k8s.io/client-go to v0.21.0-rc.0 introduced a change that broke controller-runtime, which necessitated bumping controller-runtime to v0.9.0-alpha.1.
* Bumping controller-runtime to v0.9.0-alpha.1 introduced a breaking change, which necessitated minor changes in the operator manager setup (it actually required reverting a change that had been needed for the controller-runtime v0.8.3 bump).

So this PR now includes the following commits:

* Steve's commits to bump to controller-runtime v0.8.3 plus fixups for that bump.
* Another commit to bump openshift/api and bump to k8s.io v0.21.0-rc.0 and controller-runtime v0.9.0-alpha.1 plus fixups for that bump.
* A commit that implements the unsupported config override API.

Let's see if that is enough...

**Comment by Miciah** [comment]
Rebased.

**Comment by Miciah** [comment]
/hold
for https://github.com/openshift/router/pull/192.

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -53,24 +53,6 @@ type Operator struct {
 	namespace string
 }
 
-type newClientBuilder struct{}
```
Comment (sgreene570) [inline pkg/operator/operator.go]: Ah, I figured there was a better way. Thanks for patching this!

File: test/e2e/http_header_buffer_test.go [hunk 2]
```diff
@@ -102,7 +102,7 @@ func TestHTTPHeaderBufferSize(t *testing.T) {
 	// Generate an arbitrary request header with random phony data.
 	// Use a random string with length reasonably close to the
 	// buffer limits to save space for other headers.
-	length := int(ic.Spec.HTTPHeaderBuffer.HeaderBufferBytes-ic.Spec.HTTPHeaderBuffer.HeaderBufferMaxRewriteBytes) - 1024
+	length := int(ic.Spec.TuningOptions.HeaderBufferBytes-ic.Spec.TuningOptions.HeaderBufferMaxRewriteBytes) - 1024
```
Comment (sgreene570) [inline test/e2e/http_header_buffer_test.go:105]: Thanks for addressing the `HTTPHeaderBuffer` -> `TuningOptions` change!

File: test/e2e/operator_test.go [hunk 3]
```diff
@@ -57,6 +58,11 @@ import (
 )
 
 var (
+	availableConditionsForPrivateIngressController = []operatorv1.OperatorCondition{
```
Comment (sgreene570) [inline test/e2e/operator_test.go:61]: Nice improvement.

File: test/e2e/operator_test.go [hunk 4]
```diff
@@ -1782,6 +1763,47 @@ func TestUniqueIdHeader(t *testing.T) {
...
+// TestLoadBalancingAlgorithmUnsupportedConfigOverride verifies that the
+// operator configures router pod replicas to use the "leastconn" load-balancing
+// algorithm if the ingresscontroller is so configured using an unsupported
+// config override.
+func TestLoadBalancingAlgorithmUnsupportedConfigOverride(t *testing.T) {
+	icName := types.NamespacedName{Namespace: operatorNamespace, Name: "leastconn"}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatalf("failed to create ingresscontroller: %v", err)
```
Comment (sgreene570) [inline test/e2e/operator_test.go:1775]: Do we want to start using `%w` for errors in our e2e test code?

File: pkg/operator/controller/ingress/deployment_test.go [hunk 5]
```diff
@@ -381,6 +389,11 @@ func TestDesiredRouterDeployment(t *testing.T) {
 
 	checkDeploymentHasEnvVar(t, deployment, "ROUTER_IP_V4_V6_MODE", true, "v4v6")
 
+	// Any value for loadBalancingAlgorithm other than "leastconn" should be
+	// ignored.
+	ci.Spec.UnsupportedConfigOverrides = runtime.RawExtension{
+		Raw: []byte(`{"loadBalancingAlgorithm":"random"}`),
```
Comment (sgreene570) [inline pkg/operator/controller/ingress/deployment_test.go:395]: nit: Might want to use a value other than `random` here (ie `foo` or something unrelated to `loadBalancingAlgorithm`).

File: test/e2e/operator_test.go (hunk 4 above)
Comment (Miciah) [reply test/e2e/operator_test.go:1775]: I cannot use `%w` in `t.Fatalf`:

    test/e2e/operator_test.go:1775:3: Fatalf call has error-wrapping directive %w

However, I changed the `t.Errorf` call below to use `%w`.

File: pkg/operator/controller/ingress/deployment_test.go (hunk 5 above)
Comment (Miciah) [reply pkg/operator/controller/ingress/deployment_test.go]: Fixed.  Thanks!

**Comment by sgreene570** [comment]
/lgtm
/hold cancel
Deadlines are approaching :ship:

**Comment by sgreene570** [comment]
/hold

We want to run this PR through CI once https://github.com/openshift/router/pull/192 merges

File: pkg/operator/operator.go (hunk 1 above)
Comment (Miciah) [reply pkg/operator/operator.go]: I think `ClientBuilder` was needed for v0.8.3, but then v0.9.0 went back to using `NewCacheFunc`.

Comment (sgreene570) [reply pkg/operator/operator.go]: Right, my mistake.

Either way, glad it's simplified now!

**Comment by sgreene570** [comment]
In the meantime lets make sure e2e-aws-operator works (last run was an installation fluke)
 
/test e2e-aws-operator

**Comment by sgreene570** [comment]
https://github.com/openshift/router/pull/192 has merged! 

/retest

**Comment by Miciah** [comment]
Logs from the [router-default-56f59677db-k4kkb](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/589/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1379213177102798848/artifacts/e2e-aws-operator/gather-extra/artifacts/pods/openshift-ingress_router-default-56f59677db-k4kkb_router.log), [router-default-56f59677db-q2pt8](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/589/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1379213177102798848/artifacts/e2e-aws-operator/gather-extra/artifacts/pods/openshift-ingress_router-default-56f59677db-q2pt8_router.log), and [router-default-56f59677db-vnlr6](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-ingress-operator/589/pull-ci-openshift-cluster-ingress-operator-master-e2e-aws-operator/1379213177102798848/artifacts/e2e-aws-operator/gather-extra/artifacts/pods/openshift-ingress_router-default-56f59677db-vnlr6_router.log) pods all have the following:

    template.go:433] router "msg"="starting router"  "version"="majorFromGit: \nminorFromGit: \ncommitFromGit: 88ecceb\nversionFromGit: v0.0.0-unknown\ngitTreeState: dirty\nbuildDate: 2021-04-05T19:39:50Z\n"

This means that tests passed with https://github.com/openshift/router/commit/88ecceb8953751fcde2fe2eab52a5c15e07a955e, which is the merge commit for https://github.com/openshift/router/pull/192, so CI has verified the router change along with the operator change.  

/hold cancel

**Pull request merged**